Kernel/Graphics/GraphicsManagement.cpp

```diff
@@ -140,6 +141,8 @@ UNMAP_AFTER_INIT bool GraphicsManagement::initialize()
         }
         display_adapter->initialize_framebuffer_devices();
     });
+    if (m_graphics_devices.is_empty())
+        PANIC("No graphical device was initialized.");
```

Comment:
I know that SerenityOS has an emphasis on a graphical desktop, but that's a bit too extreme. Please think about those who can't afford a graphics card during these times with the current massive shortages :^)

Reply:
it's not about graphical desktop, it's about that you just can't have anything useful with the system if you don't have a graphical device of some sort. If we didn't find any graphical device, it's most likely an error though, like what #7334 where @linusg booted his system and the kernel didn't find any VGA compatible card although he had (GeForce GT 1030).
If we don't find a graphical device of some sort, we can't initialize safely even a text console (VGA text console, or high resolution framebuffer for Framebuffer console).

Maybe later on when we have serial terminal support implemented in the kernel, we could let people to boot without any graphics adapter in their system (although this might be a rare case, because the firmware might refuse to boot without it).

Reply:
There is work on progress for serial terminal support (#7260) and a computer doesn't need a graphics card to function. Even without going into embedded or specialized devices, I could pull out the graphic card in my desktop computer (which doesn't have an iGPU) and repurpose it as a headless server if I wanted to.

The point is, x86 computers without graphic cards do exist and unless it is decided that SerenityOS will not support headless operation, then the kernel shouldn't panic if there are no graphic cards in the system.

Reply:
I've seen the work in progress of serial terminal support. I encourage to support headless operation, although this is definitely not a main goal of the project nor usual use case of our project.
Also - "Doesn't need a graphics card to function" doesn't mean "Can boot without a graphics card", if we take it from the user perspective and not from OS developer point of view, because I can simply unplug my computer monitor and then the graphics card is still connected, but the OS and the computer don't need it to function.

If the BIOS/UEFI firmware supports it, then no problem, we can support it too. I just want to see evidence for such thing, otherwise how can we know we didn't miss the card when we scanned the PCI bus?
Does your machine can be booted without any kind of graphics adapter (you said you don't have an integrated GPU)? Does the firmware not halting with some beeps?